src/Http/ReceivedAuthnRequestQueryString.php

```diff
@@ -0,0 +1,292 @@
+                'Invalid ReceivedAuthnRequest query string ("%s"): parameter "%s" not found',
+                $queryString,
+                self::PARAMETER_REQUEST
+            ));
+        }
+
+        if (base64_decode(urldecode($parameters[self::PARAMETER_REQUEST]), true) === false) {
+            throw new InvalidRequestException('Failed decoding SAML request, did not receive a valid base64 string');
+        }
+
+        $parsedQueryString = new self($parameters[self::PARAMETER_REQUEST]);
+
+        if (isset($parameters[self::PARAMETER_RELAY_STATE])) {
+            $parsedQueryString->relayState = $parameters[self::PARAMETER_RELAY_STATE];
+        }
+        if (isset($parameters[self::PARAMETER_SIGNATURE])) {
+            if (base64_decode(urldecode($parameters[self::PARAMETER_SIGNATURE]), true) === false) {
+                throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                    'Invalid ReceivedAuthnRequest query string ("%s"): signature is not base64 encoded correctly',
+                    $queryString
+                ));
+            }
+
+            $parsedQueryString->signature = $parameters[self::PARAMETER_SIGNATURE];
+        }
+        if (isset($parameters[self::PARAMETER_SIGNATURE_ALGORITHM])) {
+            $parsedQueryString->signatureAlgorithm = $parameters[self::PARAMETER_SIGNATURE_ALGORITHM];
+        }
+
+        if ($parsedQueryString->signatureAlgorithm === null && $parsedQueryString->signature !== null) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Invalid ReceivedAuthnRequest query string ("%s") contains a signature but not a signature algorithm',
+                $queryString
+            ));
+        }
+        if ($parsedQueryString->signature === null && $parsedQueryString->signatureAlgorithm !== null) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Invalid ReceivedAuthnRequest query string ("%s") contains a signature algorithm but not a signature',
+                $queryString
+            ));
+        }
+
+        return $parsedQueryString;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasRelayState()
+    {
+        return $this->relayState !== null;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSignedQueryString()
+    {
+        $query = self::PARAMETER_REQUEST . '=' . $this->samlRequest;
+
+        if ($this->hasRelayState()) {
+            $query .= '&' . self::PARAMETER_RELAY_STATE . '=' . $this->relayState;
+        }
+
+        if ($this->isSigned()) {
+            $query .= '&' . self::PARAMETER_SIGNATURE . '=' . $this->signature
+               . '&' . self::PARAMETER_SIGNATURE_ALGORITHM . '=' . $this->signatureAlgorithm;
+        }
+
+        return $query;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSignableQueryString()
```

Comment:
This method can go, there should not be a reason to look at querystring simply for signing purpose (i.e. that would be the to-be-created `ConstructedAuthnRequest`)